File: btlweb3/Scripts/Admin/HoaDonBanController.js
/// <reference path="app.js" />

app.controller("HoaDonBanController", ($scope, $rootScope, $http) => {
    $scope.rowCount = 0;
    $scope.currentPage = 1;
    $scope.pageSize = 10;
    $rootScope.ChiTiets = [];

    // lấy tất cả sản phẩm
    $http.get('/Admin/HoaDonBan/GetAllTenSanPham').then(res => {
        $rootScope.SanPhams = res.data;
    });

    // hàm lấy ra tên sp, đưa vào mã sp sẽ trả về tên sp tương ứng
    $scope.getTenSP = (MaSP) => {
        return $rootScope.SanPhams.filter(m => m.MaSP == MaSP)[0].TenSP;
    };

    // hàm lấy ra khách hàng, đưa vào mã khách hàng sẽ trả về khách hàng tương ứng
    $scope.getKhachHang = (MaKH) => {
        return $rootScope.KhachHangs.filter(m => m.MaKH == MaKH)[0];
    };

    $scope.pageChanged = () => {
        loadData();
    };

    loadData();
    function loadData() {
        $http.get('/Admin/HoaDonBan/GetAllTenSanPham').then(res => {
            $rootScope.SanPhams = res.data;
        });

        $http.get('/Admin/HoaDonBan/GetAllKhachHang').then(res => {
            $rootScope.KhachHangs = res.data;
        });

        $http.get('/Admin/HoaDonBan/GetData', {
            params: {
                page: $scope.currentPage,
                pageSize: $scope.pageSize,
            }
        }).then(res => {
            // duyệt qua các kết quả trả về
            $scope.table = res.data.Results.map(hd => {
                // set khách hàng vào các kết quả
                hd.khachHang = $scope.getKhachHang(hd.MaKH);
                return hd;
            });
            $scope.rowCount = res.data.RowCount;
        });
    }

    // hàm chạy khi ng dùng ấn nút thêm sp ở trong hóa đơn mới
    $rootScope.onAdd = () => {
        $rootScope.ChiTiets.push({
            SoLuong: 1,
        });
    }

    $scope.onCreate = () => {
        $rootScope.isSubmited = false;
        $rootScope.formModel = undefined;
        $rootScope.ChiTiets = [];
        showModal();
    }

    $scope.onDelete = (item) => {
        swAlert.delete("Bạn có thật sự muốn xóa hóa đơn này?", () => {
            $http.post('/Admin/HoaDonBan/Xoa', { id: item.MaHDB }).then(res => {
                swAlert.success("Xóa hóa đơn thành công.");
                loadData();
            });
        });
    }

    // hàm chạy khi ng dùng ấn xem 1 hóa đơn
    $scope.onView = (item) => {
        // get hóa đơn bằng mã từ controller mvc
        $http.get('/Admin/HoaDonBan/ViewHoaDon', {
            params: {
                id: item.MaHDB,
            }
        }).then(res => {
            // duyệt qua các chi tiết trong hóa đơn đc trả về
            res.data.chiTiet = res.data.chiTiet.map(ct => {
                // set giá trị cho sản phẩm trong từng chi tiết
                ct.SanPham = $scope.getSanPham(ct.MaSP);
                return ct;
            });
            // gán toàn bộ dữ liệu hóa đơn cho biến viewModel để hiển thị lên modal
            $rootScope.viewModel = res.data;
            $rootScope.ChiTiet = [];
        });
        $("#viewModal").modal("show");
    }

    $rootScope.onSubmit = () => {
        $rootScope.isSubmited = true;

        if (!$rootScope.form.$valid) {
            return;
        }

        // gửi dữ liệu lên controller mvc bao gồm thông tin khách hàng và các chi tiết sản phẩm
        $http.post('/Admin/HoaDonBan/Them', {
            KhachHang: $rootScope.formModel,
            ChiTietHDBs: $rootScope.ChiTiets,
        })
            .then(res => {
                swAlert.success("Thêm hóa đơn thành công.");
                loadData();
            });

        hideModal();
    }

    function showModal() {
        $("#myModal").modal("show");
    }

    function hideModal() {
        $("#myModal").modal("hide");
    }    
});

File: btlweb3/Areas/Admin/Controllers/HoaDonBanController.cs
using btlweb3.Helper;
using btlweb3.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Data.Entity.Migrations;
using System.IO;
using System.Web;
using System.Web.Mvc;

namespace btlweb3.Areas.Admin.Controllers
{
    public class HoaDonBanController : Controller
    {
        public Ban_GiayEntities db = new Ban_GiayEntities();

        // GET: Admin/HoaDonBan
        public ActionResult Index()
        {
            return View();
        }

        // lấy tất cả bản ghi khách hàng, k phân trang
        public JsonResult GetAllKhachHang()
        {
            db.Configuration.ProxyCreationEnabled = false;
            return Json(db.KhachHangs, JsonRequestBehavior.AllowGet);
        }

        public JsonResult GetAllTenSanPham()
        {
            db.Configuration.ProxyCreationEnabled = false;
            var list = db.SanPhams.ToList();
            return Json(db.SanPhams, JsonRequestBehavior.AllowGet);
        }

        // hàm getdata để lấy dữ liệu theo số trang mà bên angularjs gửi sang
        public JsonResult GetData(int page, int pageSize)
        {
            db.Configuration.ProxyCreationEnabled = false;
            // dùng pagination vừa tạo để phân trang
            var result = Pagination.GetPaged(db.HoaDonBans, page, pageSize);

            return Json(result, JsonRequestBehavior.AllowGet);
        }

        // lấy hóa đơn theo mã hóa đơn
        public JsonResult ViewHoaDon(int id)
        {
            db.Configuration.ProxyCreationEnabled = false;
            // lấy hóa đơn theo mã :))
            var onView = db.HoaDonBans.Where(hd => hd.MaHDB == id)
                .Select(hd => new
                {
                    hd.MaHDB,
                    hd.MaKH,
                    khachHang = db.KhachHangs.FirstOrDefault(kh => kh.MaKH == hd.MaKH),
                    chiTiet = db.ChiTietHDBs.Where(ct => ct.MaHDB == hd.MaHDB)
                }).First();

            return Json(onView, JsonRequestBehavior.AllowGet);
        }

        // hàm thêm nhận dữ liệu ng dùng nhập bên angularjs rồi thêm vào csdl
        public JsonResult Them(HoaDonBan model)
        {
            db.Configuration.ProxyCreationEnabled = false;
            db.KhachHangs.Add(model.KhachHang);
            db.SaveChanges();

            HoaDonBan hd = new HoaDonBan();
            hd.MaKH = model.KhachHang.MaKH;
            hd.NgayTao = DateTime.Now;

            db.HoaDonBans.Add(hd);
            db.SaveChanges();

            foreach (var item in model.ChiTietHDBs)
            {
                // duyệt qua giỏ hàng rồi lưu lại từng sp trong giỏ vào chi tiết của đơn hàng trên
                db.Configuration.ProxyCreationEnabled = false;
                ChiTietHDB ct = new ChiTietHDB();
                ct.MaHDB = hd.MaHDB;
                ct.SanPham.TenSP = item.SanPham.TenSP;
                ct.SoLuong = item.SoLuong;
                ct.DonGia = db.SanPhams.Find(item.MaSP).DonGia;
                db.ChiTietHDBs.Add(ct);
                db.SaveChanges();
            }

            return Json(true);
        }

        public JsonResult Xoa(int id)
        {
            db.Configuration.ProxyCreationEnabled = false;
            var model = db.HoaDonBans.Find(id);
            db.HoaDonBans.Remove(model);
            db.SaveChanges();

            return Json(model);
        }
    }
}